### ISSUE TITLE
how to improve the performance in detecting small words?

### ISSUE THREAD

**Issue by jycloud** [eragonruan/text-detection-ctpn#52]

**Comment by eragonruan** [comment]
@jycloud some potential solutions to this problem.  
- multi-scale testing, this is the most direct way.
- smaller anchor size when training
- feature fusion, since small words may disappear in feature map after several pooling op

**Comment by jycloud** [comment]
@eragonruan  
1.multi-scale testing is a simple way,but the cnn model resize input size to 1000*600,if I divide the pictures,There is a probability to cut the character,it may cause a bad result.
2.The ANCHOR_SCALES == 16,In training phase, you mean I should reset it to 14 or smaller?
3.I dont think so,I use this code to detect paper documents,my results shows:most(about 90%) small words(about 20*450pixels)could be detected,but some(about 10%) could not.

**Comment by eragonruan** [comment]
@jycloud the model can take any size of input(short size should longer than 300), 600x1000 is just a default setting I set. 
no. anchor size is 11-273

**Comment by jycloud** [comment]
@eragonruan In generate_anchors.py,I find the anchor size setting,the small words' height is more than 11, I guess the key is not anchor size,the error is like this:
![1130](https://user-images.githubusercontent.com/31645743/33417157-37566a2c-d5db-11e7-802f-dc95287da36c.png)
how can I optimize ?

**Comment by eragonruan** [comment]
how many rows do you have, I only keep 1000 proposals during test, for your case, each row 20x450 needs about 30 proposals, hence if your document has more than 33 rows, the model may miss something

**Comment by jycloud** [comment]
@eragonruan where can I modify proposals parameter?

**Comment by eragonruan** [comment]
@jycloud [lib/fast_rcnn/config.py#L179](https://github.com/eragonruan/text-detection-ctpn/blob/839aa32dec5a9c8813c595ccc7668888776150bd/lib/fast_rcnn/config.py#L179)

**Comment by jycloud** [comment]
@eragonruan I reset the parameter to 2000,but results did not change.

**Comment by jycloud** [comment]
@eragonruan I decide to retrain this model by my own dataset,hoped I can solve this problem,by the way,in training phase,the default setting is resize pictures to (1000,600) or (600,1000)?

**Comment by eragonruan** [comment]
@jycloud short side is resized to 600 pixel. afer resize, if long side is longer than 1200 pixel, resize the long side to 1200 pixel.

**Comment by jycloud** [comment]
@eragonruan Thank you very much.

**Comment by guddulrk** [comment]
@eragonruan where can I modify iterations (150000)? I am training the model on CPU, so it is taking too much time. I want to reduce the iterations..

Thanks

**Comment by eragonruan** [comment]
@guddulrk [ctpn/text.yml#L10](https://github.com/eragonruan/text-detection-ctpn/blob/5d091787c9ba2e0c5da919341a9fe11117cceab2/ctpn/text.yml#L10)

**Comment by guddulrk** [comment]
Thanks, mate.

**Comment by guddulrk** [comment]
@eragonruan should I change stepsize and snapshot_iter? as there is SOLVER: Adam on #L10.

restore: 0
  SOLVER: Adam
  OHEM: False
  RPN_BATCHSIZE: 300
  BATCH_SIZE: 300
  LOG_IMAGE_ITERS: 100
  DISPLAY: 10
  SNAPSHOT_ITERS: 1000
  HAS_RPN: True
  LEARNING_RATE: 0.00001
  MOMENTUM: 0.9
  GAMMA: 0.1
  STEPSIZE: 70000
  IMS_PER_BATCH: 1

**Comment by eragonruan** [comment]
@guddulrk  use the latest code

**Comment by guddulrk** [comment]
@eragonruan thank you so much.

**Comment by guddulrk** [comment]
@eragonruan I am training the model on new dataset but getting nan values as shown below:
speed: 17.140s / iter
iter: 3390 / 5000, total loss: nan, model loss: nan, rpn_loss_cls: nan, rpn_loss_box: nan, lr: 0.000010

can you please help me to sort out the issue?
Thanks

**Comment by eragonruan** [comment]
@guddulrk this may caused by your training data, check [here](https://github.com/rbgirshick/py-faster-rcnn/issues/159) for more detail.

**Comment by kyosocan** [comment]
@eragonruan Thanks for your code. 
Now, I'm trying to change the anchor's width from 16 to 10, should I change the parameter _feat_stride as well? 
Looking forward to your reply, thank you!

**Comment by eragonruan** [comment]
@kyosocan I don't think 10 is an option. the anchor's width is decided by the feature layer you choose. 
for VGG net, conv5_3==>16, conv4_3==>8

**Comment by kyosocan** [comment]
@eragonruan 　Thank you very much for your reply. Now I see how to adjust this parameter.  In addition, if   
the resnet is better than VGG 16 in this case ?

**Comment by eragonruan** [comment]
@kyosocan sorry, I did not try model with resnet. but it's worth trying.
